agent/taskresource/envFiles/envfile.go

```diff
@@ -0,0 +1,432 @@
+	// so this should be ok to be called by multiple go routines
+	tmpFile, err := envfile.ioutil.TempFile(envfile.resourceDir, envTempFilePrefix)
+	if err != nil {
+		seelog.Infof("something went wrong trying to create a temp file with prefix %s", envTempFilePrefix)
+		return err
+	}
+	defer tmpFile.Close()
+
+	err = writeFunc(tmpFile)
+	if err != nil {
+		seelog.Infof("something went wrong trying to write to tmpFile %s", tmpFile.Name())
+		return err
+	}
+
+	// persist file to disk
+	err = tmpFile.Sync()
+	if err != nil {
+		seelog.Infof("something went wrong trying to persist envfile to disk")
+		return err
+	}
+
+	err = envfile.os.Rename(tmpFile.Name(), fullPathName)
+	if err != nil {
+		seelog.Infof("something went wrong when trying to rename envfile from %s to %s", tmpFile.Name(), fullPathName)
+		return err
+	}
+
+	return nil
+}
+
+func (envfile *EnvironmentFileResource) getEnvironmentFilesSource() []apicontainer.EnvironmentFile {
+	envfile.lock.RLock()
+	defer envfile.lock.RUnlock()
+
+	return envfile.environmentFilesSourceUnsafe
+}
+
+func (envfile *EnvironmentFileResource) getExecutionCredentialsID() string {
+	envfile.lock.RLock()
+	defer envfile.lock.RUnlock()
+
+	return envfile.executionCredentialsID
+}
+
+// Cleanup removes env files downloaded for the task
+func (envfile *EnvironmentFileResource) Cleanup() error {
+	// TODO
+	return nil
+}
+
+type EnvironmentFileResourceJSON struct {
+	TaskARN                string                         `json:"taskARN"`
+	CreatedAt              *time.Time                     `json:"createdAt,omitempty"`
+	DesiredStatus          *EnvironmentFileStatus         `json:"desiredStatus"`
+	KnownStatus            *EnvironmentFileStatus         `json:"knownStatus"`
+	EnvironmentFilesSource []apicontainer.EnvironmentFile `json:"envfileResources"`
+	ExecutionCredentialsID string                         `json:"executionCredentialsID"`
+}
+
+// MarshalJSON serializes the EnvironmentFileResource struct to JSON
+func (envfile *EnvironmentFileResource) MarshalJSON() ([]byte, error) {
+	if envfile == nil {
+		return nil, errors.New("envfile resource is nil")
+	}
+	createdAt := envfile.GetCreatedAt()
+	return json.Marshal(EnvironmentFileResourceJSON{
+		TaskARN:   envfile.taskARN,
+		CreatedAt: &createdAt,
+		DesiredStatus: func() *EnvironmentFileStatus {
+			desiredState := envfile.GetDesiredStatus()
+			envfileStatus := EnvironmentFileStatus(desiredState)
+			return &envfileStatus
+		}(),
+		KnownStatus: func() *EnvironmentFileStatus {
+			knownState := envfile.GetKnownStatus()
+			envfileStatus := EnvironmentFileStatus(knownState)
+			return &envfileStatus
+		}(),
+		EnvironmentFilesSource: envfile.getEnvironmentFilesSource(),
+		ExecutionCredentialsID: envfile.getExecutionCredentialsID(),
+	})
+
+}
+
+// UnmarshalJSON deserializes the raw JSON to an EnvironmentFileResource struct
+func (envfile *EnvironmentFileResource) UnmarshalJSON(b []byte) error {
```

Comment:
I think you miss the assignment of executionCredentialsID here.

Reply:
oops you're right, also task arn. thanks for catching